Add missed blocks to monitored validators

v4lproik opened the pull request:
Solving #3414 

Will add Prom metrics + tests tomorrow

v4lproik committed:
add missed_block metric

v4lproik committed:
init missed_block in constructor

v4lproik committed:
declare beaconproposercache in ValidatorMonitor

v4lproik committed:
refacto proposer_shuffling_decision_root to use epoch instead of current.epoch

v4lproik committed:
imple new proposer_shuffling_decision_root in callers

v4lproik committed:
push missed_blocks

v4lproik committed:
prune missed_blocks

v4lproik committed:
only add to hashmap if it's a monitored validator

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/sigp/lighthouse?pullRequest=4731) <br/>All committers have signed the CLA.

v4lproik committed:
remove current_epoch dup + typos

v4lproik committed:
extract in func

v4lproik committed:
add prom metrics

v4lproik committed:
checkpoint is not only epoch but slot as well

v4lproik committed:
add safeguard if we start a new chain at slot 0

v4lproik committed:
clean

v4lproik commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                                .into_iter()
+                                .find(|validator| validator.index == Some(proposer.index as u64))
+                                .is_some() {
+                                self.missed_blocks.insert((current_epoch, proposer.index as u64, slot));
```
confirm with the team if they want to insert the slot number or the slot % 32 of the epoch in the missed blocks hashset

michaelsproul replied:
`slot` is best

v4lproik commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                    if let Ok(shuffling_decision_block) = state.proposer_shuffling_decision_root(epoch, *block_root) {
+                        // TODO: AI(Joel) Ask the team if we can use the beacon proposer cache without the mutex
+                        // mutex is implemented in the caller (BeaconChain), not in the cache implementation
+                        if let Some(proposer) = self.beacon_proposer_cache.get_slot::<T>(shuffling_decision_block, slot) {
```
Ask the team about the cache.

It seems that the cache is used with a mutex in the BeaconChain. I am not sure if:
- We also want to add a mutex here when reading the cache? (why? is the internal state of the cache changes when reading which which would require a lock?)
- Do we want a shared mutex then?

Would love some pointers here!

michaelsproul replied:
I think we probably want to share the cache with the beacon chain, which could be done by using an `Arc<Mutex<BeaconProposerCache>>` everywhere. That way the `BeaconChain` and `ValdiatorMonitor` will use the same cache (this will require a bit of plumbing after updating the definition of `beacon_proposer_cache:` in the `BeaconChain` struct).

Seeing as the validator monitor is less important than other parts of beacon chain processing we might also want to avoid locking in this loop. We could probably do 1-2 calls to `BeaconProposerCache::get_epoch` instead, and then use those values in the loop.

michaelsproul commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+            let slot = state.slot() - Slot::new(n as u64);
+            let prev_slot = slot - 1;
+
+            // safeguard if we are on a new chain, the slot - 1 can be negative
+            if prev_slot < 0 {
+                break;
+            }
```
Slot arithmetic saturates, and uses an unsigned int, so this check isn't required. Instead I think we should `continue` if `slot == 0`, because the genesis block can't be missed, and that's where we get into the weird case of `0 - 1 == 0`.

<details>
the impl of saturating sub for minus is here:

https://github.com/sigp/lighthouse/blob/2841f60686d642fcc0785c884d43e34e47a800dc/consensus/types/src/slot_epoch_macros.rs#L137-L143

in some places (consensus code) we enforce that the implicit saturating operators aren't used, and require explicit `saturating_sub` or `safe_sub`
</details>

michaelsproul commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+    fn add_validators_missed_blocks(&mut self, current_epoch: Epoch, state: &BeaconState<T>) {
+        // Determine missed (non-finalized) blocks (can contain false positives)
+        let range_of_slot = (T::slots_per_epoch() as usize - MISSED_BLOCK_LAG_SLOTS) - 1;
+        for n in 0..range_of_slot {
+            let slot = state.slot() - Slot::new(n as u64);
+            let prev_slot = slot - 1;
```
I think for this range we want something like:

- start: `state.slot() - MISSED_BLOCK_LAG_SLOTS - T::slots_per_epoch()`
- end: `state.slot() - MISSED_BLOCK_LAG_SLOTS`

i.e. a range of length `slots_per_epoch` that ends 4 blocks before the state's current slot. At the moment the loop will start at `state.slot()` (when `n=0`), which negates the effect of the buffer -- we want to wait for a few slots to pass before declaring a block missed.

michaelsproul commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                break;
+            }
+
+            // condition for missed_block is defined such as block_root == block_root - n
+            // where the proposer who missed the block is the proposer of the block at block_root - n
```
This code is out of sync with the code (which is correct). The proposer who missed is the one at `slot` where `block_root(slot) == block_root(slot - 1)`.

michaelsproul commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+            if let (Ok(block_root), Ok(prev_block_root)) = (state.get_block_root(slot), state.get_block_root(prev_slot)) {
+                if block_root == prev_block_root {
+                    let epoch = slot.epoch(T::slots_per_epoch());
+                    if let Ok(shuffling_decision_block) = state.proposer_shuffling_decision_root(epoch, *block_root) {
```
I think seeing as we're skipping the genesis block we could just pass `Hash256::zero()` for the block root here (it won't be used, see the docs for `proposer_shuffling_decision_root`).

michaelsproul commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                                .into_iter()
+                                .find(|validator| validator.index == Some(proposer.index as u64))
+                                .is_some() {
+                                self.missed_blocks.insert((current_epoch, proposer.index as u64, slot));
```
I think this should be the epoch of the proposal, which also means we don't need to pass in `current_epoch`.

```suggestion
                                self.missed_blocks.insert((epoch, proposer.index as u64, slot));
```

michaelsproul commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                                self.log,
+                                "Could not find proposer for a missed non-finalized block in beacon proposer cache";
+                                "slot" => slot,
+                                "block_root" => format!("{}", block_root),
+                                "shuffling_decision_block" => format!("{}", shuffling_decision_block),
```
`slog` has a little signifier for using the Display impl:

```suggestion
                                "block_root" => %block_root,
                                "shuffling_decision_block" => %shuffling_decision_block,
```

you can also get Debug, which would show the full block root, with `?name`

michaelsproul commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -487,8 +500,66 @@ impl<T: EthSpec> ValidatorMonitor<T> {
                     );
                 }
+
+                // Add non finalized missed blocks for the monitored validators for the current slot of the current epoch
+                let current_slot = state.slot();
```
With the lag implemented this won't catch any missed blocks by checking at `state.slot()`. I think instead we could just separately iterate `missed_blocks` and throw them all into the metric. The metric type could also be changed to an `int_gauge_vec` so that it doesn't increment multiple times for the same missed block.

see: https://prometheus.io/docs/concepts/metric_types/

michaelsproul reviewed (commented):
looking good! thanks for digging into this!

I've spammed a bunch of comments, hopefully they're helpful

v4lproik committed:
remove unnecessary negative value for a slot

v4lproik committed:
typo in comment

v4lproik committed:
remove unused current_epoch

v4lproik committed:
share beacon_proposer_cache between validator_monitor and beacon_chain

v4lproik committed:
pass Hash256::zero()

v4lproik committed:
debug objects

v4lproik committed:
fix loop: lag is at the head

v4lproik committed:
sed s/get_slot/get_epoch

v4lproik committed:
fewer calls to cache.get_epoch

v4lproik committed:
fix typos

v4lproik committed:
remove cache first call

v4lproik committed:
export TYPICAL_SLOTS_PER_EPOCH and use it in validator_monitor

v4lproik committed:
switch to gauge & loop over missed_blocks hashset

v4lproik committed:
fix subnet_service tests

v4lproik commented:
Thanks for all the comments! They were extremely useful! I have started taking into account all your comments and I will finalize the PR with some tests very shortly.

v4lproik committed:
remove unused var

v4lproik committed:
clean + fix nits

v4lproik committed:
add beacon_proposer_cache + validator_monitor in builder

v4lproik committed:
fix store_tests

v4lproik committed:
fix builder tests

v4lproik committed:
add tests

v4lproik committed:
add validator monitor set of tests

v4lproik committed:
clean tests

v4lproik committed:
nits

v4lproik committed:
optimise imports

v4lproik committed:
lint

v4lproik commented:
I allowed myself to refactor the BeaconChainHarness builder so I could mock the beacon_proposer_cache as well as passing an instance of validator_monitor so I could easily change the monitored validators list.

I will be adding the graph representing the missed block metric in the lighthouse-metrics and open another PR this week adding the missed finalized blocks.

As a side note, it was very fun spending some time in the beacon chain and read its implementation alongside the eth2book :)

v4lproik commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -37,6 +40,9 @@ pub const HISTORIC_EPOCHS: usize = 10;
 pub const DEFAULT_INDIVIDUAL_TRACKING_THRESHOLD: usize = 64;
 
+/// Lag slots used in detecting missed blocks for the monitored validators
+pub const MISSED_BLOCK_LAG_SLOTS: usize = 4;
```
Not a big fan of a const here. If we are using MinimalEthSpec for instance considering that an epoch consists of 8 slots, I don't think we want the lag to be at 4. We might want to be different depending on the spec used. Wdyt?

paulhauner replied:
I feel you, but I think it's OK to assume 32 slots per epoch here. All our meaningful testnets are set to 32 and I'd still probabaly lean towards a value of 4 on the minimal spec too.

So, I'd vote to leave this as it.

paulhauner commented on consensus/types/src/beacon_state.rs:
```diff
@@ -631,8 +632,8 @@ impl<T: EthSpec> BeaconState<T> {
 
     /// Returns the slot at which the proposer shuffling was decided. The block root at this slot
     /// can be used to key the proposer shuffling for the current epoch.
```
```suggestion
    /// can be used to key the proposer shuffling for the given epoch.
```

Just a minor nit!

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +505,88 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+            .into_iter()
+            .for_each(|(validator_index, group)| {
+                let missed_blocks_count = group.count();
+                metrics::set_int_gauge(
```
The validator manager has a concept of enabling/disabling "individual tracking". We had to introduce this because nodes that are tracking hundreds/thousands of validators end up with a Prometheus metrics cardinality problem. Sigma Prime has some BNs running thousands of validators on testnets and the resulting metrics output was grinding our Prometheus server to a halt.

To resolve this we switched to *optionally* tracking each individual validator (disabled by default for > 64 validators) and *always* tracking the aggregate of all validators (see the `TOTAL_LABEL`).

To this end, would you mind gating the per-validator metrics behind `self.individual_tracking()` and also exposing an aggregate under `TOTAL_LABEL`?

v4lproik replied:
That makes sense! I used the already established self.aggregatable_metric to do that.

paulhauner commented on consensus/types/src/beacon_state.rs:
```diff
@@ -613,15 +613,16 @@ impl<T: EthSpec> BeaconState<T> {
-    pub fn proposer_shuffling_decision_root(&self, block_root: Hash256) -> Result<Hash256, Error> {
-        let decision_slot = self.proposer_shuffling_decision_slot();
+    /// The `epoch` should be set to the epoch of the block root.
+    pub fn proposer_shuffling_decision_root(&self, epoch: Epoch, block_root: Hash256) -> Result<Hash256, Error> {
```
I think we need to create a new `proposer_shuffling_decision_root_at_epoch` function rather than modifying this one.

The issue with modifying this one is that it becomes possible to provide an `epoch` greater than `self.current_epoch()`. In that case we need `block_root` to be the latest block applied to `self`, which means it's no longer valid to provide the genesis block root (which is something we rely upon elsehwere in the code).

The new function implementation should be identical to this one, except we modify `self.slot() == decision_root` to use `<=` rather than `==`. We can leave `proposer_shuffling_decision_slot` as you've modified it and use it for both functions.

paulhauner commented on beacon_node/beacon_chain/src/builder.rs:
```diff
@@ -606,6 +609,22 @@ where
         self
     }
 
+    pub fn beacon_proposer_cache(
```
I don't think there's ever a case where we want to build a `BeaconChain` without a proposer cache, so it might be simpler if we just always build it in `Self::build` if it wasn't already built by something else (e.g., `Self::monitor_validators)`. That would remove the need for always calling the `self.beacon_proposer_cache` when building.

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +505,88 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                            .and_then(|proposers| proposers.get(slot_in_epoch.as_usize())) {
+                            if self.validators
+                                .values()
+                                .filter_map(|v| v.index)
```
Rather than iterating `self.validators` we could use `self.indices` to resolve the index to a pubkey and then use `self.validators.get()`. That should reduce the complexity from `O(self.validators.len())` to `O(1)`. My suggestion is likely slower in practice for smaller validator counts but it might come in useful for nodes that are monitoring hundreds/thousands of validators.

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +505,88 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+
+        // Prune missed blocks that are prior to last finalized epoch
+        let finalized_epoch = state.finalized_checkpoint().epoch;
+        self.missed_blocks.retain(|(epoch, _, _)| *epoch >= finalized_epoch);
```
With the current pruning method I think we're going to see the count of missed blocks decrease when an epoch is finalized. This might prove challenging for users to track in the case where the metric might simultaneously decrease due to pruning whilst also increasing due to a new missed block. In that case they'd see no change in the metric.

Perhaps we'd be best to track also track a never-decreasing "total missed blocks" value per validator which we can use to ensure that the metrics are never-decreasing.

paulhauner reviewed (changes requested):
Heyo, this is looking really good! I can see you've really got the gist of this and I love the comprehensive tests.

My apologies for the delays here, I took a couple of weeks of leave and I'm only just catching up.

I've left some comments here, I don't think there's anything particularly major.

v4lproik committed:
typo

v4lproik committed:
added self.aggregatable

v4lproik committed:
duplicate proposer_shuffling_decision_root

v4lproik committed:
remove duplication in passing beacon_proposer_cache

v4lproik committed:
remove duplication in passing beacon_proposer_cache

v4lproik committed:
using indices

v4lproik committed:
fmt

v4lproik committed:
implement missed blocks total

v4lproik committed:
nits

v4lproik commented:
@paulhauner I took into account your feedbacks. Let me know what you think.

Also, I would love to get on with analysing of the [BeaconProcessor](https://github.com/eth-protocol-fellows/cohort-four/blob/master/projects/project-ideas.md#by-paul-hauner) analysis and also with simulating more than 1 attestion per epoch and see what can be done! Let me know if you have any pointers! Thanks!

paulhauner commented on beacon_node/beacon_chain/src/lib.rs:
```diff
@@ -1,3 +1,4 @@
+#![recursion_limit = "256"]
```
We can avoid changing the recursion limit for the whole crate by moving the new metrics into a different `lazy_static` block. I would suggest moving them into the bottom of this block: https://github.com/v4lproik/lighthouse/blob/dc19c4c5cc6b627f77d7460c890ab5d761ecfcde/beacon_node/beacon_chain/src/metrics.rs#L995

paulhauner committed:
Fix valdiator monitor builder pattern

Unify validator monitor config struct

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +509,131 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+            });
+
+        // add to the total missed blocks by monitored validator
+        self.total_missed_blocks
```
The `to_owned` here is cloning the `total_missed_blocks` hash map, which costs heap allocations.

We can avoid this be replacing the `.to_owned().into_iter()` with `.iter()` as long as we dereference `c` (e.g. `*c as i64`).

paulhauner reviewed (changes requested):
Heyo, I started a review but then realised there are some lint issues. You can run `make lint` locally to find them :)

I realised that the `BeaconChainBuilder` is pretty wonky when it comes to the validator manager (this is nothing to do with you). It made me a little concerned that we might accidentally create two different instances of the `BeaconProposerCache` in the future. I fixed the issues with the builder over in [this commit](https://github.com/paulhauner/lighthouse/commit/13b73d6bc49098390a91fdee8f3cc78b998100f4), please feel free to cherry-pick it into this branch 🙏 

I'll come back for another round of review once the lints are addressed!

v4lproik committed:
avoid heap allocation

v4lproik committed:
remove recursion limit

v4lproik committed:
fix lint

v4lproik commented:
My bad for forgetting the linter, I'll make sure to not forget next time! Sorry about that 🙏🏼 Linter + BN tests are green.

I have updated the PR with your feedbacks and added your commit.

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,130 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                                        i,
+                                        self.total_missed_blocks.get(&i).unwrap_or(&0) + 1,
+                                    );
+                                }
```
```suggestion
                                } else {
                                    warn!(
                                        self.log,
                                        "Missing validator index";
                                        "info" => "potentially inconsistency in the validator manager",
                                        "index" => i,
                                    )
                                }
```

I don't think we'd hit this, but it would be nice to know if we have a bug here.

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,130 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                                }
+                            }
+                        } else {
+                            debug!(
+                                self.log,
+                                "Could not get proposers for epoch {slot_epoch:?} from cache"
+                            );
```
```suggestion
                            debug!(
                                self.log,
                                "Could not get proposers for from cache";
                                "epoch" => ?slot_epoch
                            );
```

This is more consistent with our logging elsewhere.

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,130 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+
+        // List of proposers per epoch from the beacon_proposer_cache
+        let mut proposers_per_epoch: Option<SmallVec<[usize; TYPICAL_SLOTS_PER_EPOCH]>> = None;
+        for n in 1..range_of_slots {
```
I'm unsure of the math in this loop. I made a sketch of it in this [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=12f71d5ef23897318364161e0bc23d15) and it's only visiting 23 slots when I'd expect it to visit 28 (`slots_per_epoch - missed_block_lag_slots`).

I've run out of time to suggest an alterative, would you mind having a look please?

v4lproik replied:
I changed the loop as one of the edge cases would make me do some if/else to dertemine whether it should get inside the loop or not. Should be more elegant that way. Now, it's going top to bottom. I used your playground [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6c58503152a5402958ea1d136d8e56e5) with different scenarios.

paulhauner commented on beacon_node/beacon_chain/src/metrics.rs:
```diff
@@ -1009,6 +1009,16 @@ lazy_static! {
+        "Number of non-finalized blocks missed at non-finalized epochs",
+        &["src", "validator"]
+    );
+    pub static ref VALIDATOR_MONITOR_MISSED_NON_FINALIZED_BLOCKS_TOTAL: Result<IntGaugeVec> = try_create_int_gauge_vec(
```
Since this is an always increasing value, the ideal Prometheus data type is an `int_counter`.

v4lproik replied:
I decided to going for a gauge as a counter can only be incremented and not set to a specific value.

In the case where we'd want to change to a counter, we'd need to prune the entire Hashmap once the Prometheus metric is updated. Let me know what's best :)

paulhauner replied:
Oh yep, makes sense! Sorry, I must have been confused!

paulhauner reviewed (commented):
Nice! No worries about missing the lints!

I've made a couple of suggestions when you have a chance :)

Regarding naming of metrics, could I suggest the following renames:

- `validator_monitor_missed_non_finalized_blocks_at_non_finalized_epoch_total`
  - Rename to: `validator_monitor_missed_non_finalized_blocks_total`
- `validator_monitor_missed_non_finalized_blocks_total`
  - Rename to: `validator_monitor_missed_blocks_total`

I'm just thinking that might be a little easier for users to read.

This is getting very close, I'm looking forward to merging it! Thanks again for your contribution here.

v4lproik committed:
renaming metrics

v4lproik committed:
renaming metrics in validator monitor

v4lproik committed:
add log if there's a missing validator index

v4lproik committed:
consistent log

v4lproik committed:
fix loop

v4lproik commented:
We are getting there! Thanks for all the help!

I was thinking about compiling my own version of LH this afternoon and run it on Holesky or Sepolia and hopefully see some metrics.

I played around with the Grafana validator monitor dashboard and added a new category called "Missed blocks" but maybe you already have an idea regarding where you'd like the metrics to be.

paulhauner commented:
> We are getting there! Thanks for all the help!

Indeed we are! You're welcome, thanks for getting involved.

> I was thinking about compiling my own version of LH this afternoon and run it on Holesky or Sepolia and hopefully see some metrics.

Once this is ready to merge I'll run it on our infra too and show you some screenshots!

> I played around with the Grafana validator monitor dashboard and added a new category called "Missed blocks" but maybe you already have an idea regarding where you'd like the metrics to be.

Nice! I have no opinions on where it should go. That new category name sounds good to me 🚀 

Regarding the loop, I have another suggestion at this [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6a1c766f61c6d2badf7da916bbea5a7f). This one uses `tuple_windows` from itertools to give us `prev_slot` and `slot`.  I make this suggestion for the following reasons:

- It allows us to avoid the `slot == 0` check. (Also, did you mean to use `continue` there rather than `break`?)
- We avoid `let range_of_slots = T::slots_per_epoch() as usize - MISSED_BLOCK_LAG_SLOTS;` which is *technically* an underflowing subtraction (but only if constants are set wrong, so it's not a big deal).

Let me know what you think! 🙏

v4lproik committed:
better loop

v4lproik committed:
move gauge to counter

v4lproik commented:
> Regarding the loop, I have another suggestion at this [Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6a1c766f61c6d2badf7da916bbea5a7f). This one uses `tuple_windows` from itertools to give us `prev_slot` and `slot`. I make this suggestion for the following reasons:

It looks better that way indeed! I was thinking about implementing the trait Iterator for the type `Slot` in order to avoid  to convert the u64 to a `Slot` in the loop right below after converting the `start_slot` and `end_slot` into u64:
```
for (prev_slot, slot) in (start_slot..=end_slot)
            .map(Slot::new)
            .tuple_windows() {  
```
However, after implementing the `func next(&mut self)` the compiler seems to be priorising the `cmp` function issued with the trait Iterator over the `cmp` from Ord resulting in compiler error at several locations in the code. This would result in rewriting `slot.cmp` into `Ord::cmp(&block.slot, x)`. Let me know if you'd rather want me to implement it.

> It allows us to avoid the slot == 0 check. (Also, did you mean to use continue there rather than break?)

I actually meant break as with the former implementation I was going through the epochs in an descending order meaning that if I had hit 0, it'd mean it's the genesis and there's no other block prior to that one.

Also, what do you think about adding a log when there's a missed block such as described for other features of the validator-monitor in [the documentation](https://lighthouse-book.sigmaprime.io/validator-monitoring.html#example-2). As a node operator I feel like it's always nice to have that log :)

paulhauner commented:
> Also, what do you think about adding a log when there's a missed block such as described for other features of the validator-monitor in [the documentation](https://lighthouse-book.sigmaprime.io/validator-monitoring.html#example-2). As a node operator I feel like it's always nice to have that log :)

Great idea! I think a log is definitely the way to go . Let's go with an `ERRO` log. I think it would be great if the log could include the slot and the root of the parent block (i.e. the duplicate root in state.block_roots). Adding the parent root will help in debugging.

v4lproik committed:
fmt

v4lproik committed:
add error message

v4lproik committed:
lint

v4lproik commented:
Done! I am rolling out a LH version with those modifications today - I wanted to rollout this on Holesky but it seems that it's still a bit of rocky boat on our side and we don't want to add too much noise if there's any issue related to those modifications. Will send you the output then 🙏🏼

v4lproik commented:
Sorry, I am being a little late with the screenshots. I'll be deploying tomorrow morning to Holesky (we had some delays internally. We do validate quite a few blocks per hour on Holesky so it'll be easier to have some data with some nice screenshots.

v4lproik committed:
fix prom metrics

v4lproik committed:
set gauge to 0 when non-finalized epochs

v4lproik commented:
I changed a few things as I wasn't setting the gauge to 0 which would result in a dashboard that wouldn't make much sense.

Now, as you can see on the screenshots below:
- the first chart displays the **total** missed (flagged at non-finalized) blocks per validator.
- the second chart displays the missed blocks during non-finalized epochs per validator.
- the third chart displays the missed blocks seen during non-finalized epochs per validator within time.

**With data**
<img width="1949" alt="Screenshot 2023-10-12 at 21 16 39" src="https://github.com/sigp/lighthouse/assets/2066462/ea0c79ba-cbd8-4d8d-9cb2-5dacd1c4e42d">

<img width="1451" alt="Screenshot 2023-10-12 at 21 19 58" src="https://github.com/sigp/lighthouse/assets/2066462/cb79ca05-5f35-4716-bff3-23ae0c926fa9">

**When no data**
<img width="1944" alt="Screenshot 2023-10-12 at 21 21 08" src="https://github.com/sigp/lighthouse/assets/2066462/e30ab40e-0513-41b7-83c9-fec150dc65ee">

**Logs**
`2023-10-12 15:08:25.040 | stderr | Oct 12 14:08:25.039 ERRO Validator missed a block                parent block root: 0xde65f9a789365650f8d6e5da86afe9ea864768008a443910be42fba7dafd51de, slot: 101438, index: 539299, service: val_mon, service: beacon`


Let me know what you think once you are done with the final review of this PR and the looks of the charts as well so I can open a PR to lighthouse-metrics. Thanks!

v4lproik committed:
better wording

paulhauner commented on consensus/types/src/beacon_state.rs:
```diff
@@ -613,6 +613,27 @@ impl<T: EthSpec> BeaconState<T> {
+    ///
+    /// The `block_root` covers the one-off scenario where the genesis block decides its own
+    /// shuffling. It should be set to the latest block applied to `self` or the genesis block root.
+    /// The `epoch` should be set to the epoch of the block root.
```
```suggestion
    /// The `block_root` must be equal to the latest block applied to `self`.
```

Since we can provide an `epoch` greater than `self.current_epoch()` the `block_root` is used for scenarios other than just the genesis "one-off scenario".

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+        validators_missed_blocks
+            .iter()
+            .for_each(|(validator_index, missed_blocks_count)| {
+                self.aggregatable_metric(validator_index.to_string().as_str(), |label| {
```
The `aggregatable_metric` function doesn't work well for gauges, unfortunately. What's going to happen here is that the `total` label is going to be set to each individual validators value and remain at the value of the last individual validator. It won't reflect the sum of all the validators.

What we'd need here is to just set the individual validator labels in this `iter` (the entire `iter` should be gated by `self.individual_tracking() == true`). Then, separately, sum the values of `validators_missed_blocks` and set the gauge for the `total` label to that.

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+            });
+
+        // Increment the Prometheus metrics counter for each missed block
+        if let Some(i) = self.last_missed_block_validator {
```
I think the `last_missed_block_validator` has a two issues:

1. It's possible that we can process a state where more than one of our validators have missed a slot. For example, there might be two skipped slots prior to some block and we might be monitoring both of the validators who missed those slots. In this case, `last_missed_block_validator` would contain the second missed proposal but not the first.
1. It's possible that we can process two different states which include the same skipped slot (see example below). In this case `last_missed_block_validator` would register the skipped slot twice.

I'll make another comment further down describing what I think is an easy fix.

### Example of duplicate skipped slots between two blocks

Consider the following block graph:

```
|S0|   |S1|   |S2|    |S3|

  |<---[❌]<--[❌]<--[✅]
  |
[✅]
  |
  |<---[❌]<--[✅]

|S0|   |S1|   |S2|    |S3|
```

- Blocks are represented by ✅ and missed blocks are ❌.
- There are two competing chains. They both agree on slot 0 and 1, but they disagree on slot 2.
- The top chain has missed blocks at slots 1 and 2.
- The bottom chain has missed blocks at only slot 1.
- When importing each of the competing chains we'll detect the missed slot 1.

v4lproik replied:
That makes great sense! Thank for the clear explanation!

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                            if let Some(pub_key) = self.indices.get(&i) {
+                                if self.validators.get(pub_key).is_some() {
+                                    // Add to missed blocks
+                                    self.missed_blocks.insert((slot_epoch, i, slot));
```
Instead of using the `last_missed_block_validator` concept, I would do something like:

```rust
if self.missed_blocks.insert((slot_epoch, i, slot)).is_none() {
  self.aggregatable_metric(i.to_string().as_str(), |label| {
    metrics::inc_counter_vec(&metrics::VALIDATOR_MONITOR_MISSED_BLOCKS_TOTAL, &[label]);
  });
  // Also, the error log about a missed slot.
}
```

[`HashMap::insert`](https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.insert) returning `None` indicates that we haven't observed this skipped slot for the given validator before. That does the metrics de-duplication for us.

This does add an edge case though! We're pruning `self.missed_blocks` at the finalised epoch and it's possible that our search back through the last 32 slots might go back into the finalised territory that we've already pruned from `self.missed_blocks` 🙀 We can fix this by pruning everything prior to `finalized_epoch - 1` rather than just `finalized_epoch`. That will give us another epoch of history to work with.

To make things a bit more future-proof with this extra pruning history, I'd suggest defining a `const MISSED_BLOCK_LOOKBACK_EPOCHS = 1`. Then you can use that value for subtracting from `finalized_epoch` *and* for defining `start_slot` on L592.

Perhaps you could add a comment to the definition of `MISSED_BLOCK_LOOKBACK_EPOCHS `  making it clear that setting it to anything higher than `1` is likely going to be problematic because the `beacon_proposer_cache` will be unlikely to contain shuffling for older epochs.

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+        // Prune missed blocks that are prior to last finalized epochs
+        let finalized_epoch = state.finalized_checkpoint().epoch;
+        self.missed_blocks
+            .retain(|(epoch, _, _)| *epoch > finalized_epoch);
```
In a future comment I'll suggest changing this pruning logic. I'm making a note here so I don't forget to check it in the next review ☺️

paulhauner commented on beacon_node/beacon_chain/src/validator_monitor.rs:
```diff
@@ -489,6 +532,152 @@ impl<T: EthSpec> ValidatorMonitor<T> {
+                    if let Ok(shuffling_decision_block) = state
+                        .proposer_shuffling_decision_root_at_epoch(
+                            slot_epoch,
+                            shuffling_decision_block,
```
I don't think it will make a practical difference, but I think we should use `block_root` rather than `Hash256::zero()` here.

`block_root` is the most correct value, but I don't think we'll ever hit the case where we need that value. I think it's worth doing though, especially since it means we can get rid of `let shuffling_decision_block = Hash256::zero();`.

paulhauner reviewed (changes requested):
Looking good! I have a few more suggestions please 🙏 

We're getting so close, I think we'll be ready to merge after this round! This is a pretty tricky component so I'm not surprised we're doing a few rounds of review.

The screenshots look great btw, I appreciate you going to the effort to test empirically and present those.

v4lproik committed:
remove hash256::zero in favour of block_root

v4lproik committed:
fix gauge total label

v4lproik committed:
fix last missed block validator

v4lproik commented:
I left your feedback about the prunning so we can verify once you look at the modifications. I am retaining one more epoch in the cache so I can correctly set the gauge with the label total as suggested. 

Also, I changed the start_slot  to 
```
let start_slot = current_epoch.start_slot(T::slots_per_epoch())
            - Slot::new(MISSED_BLOCK_LOOKBACK_EPOCHS as u64 * T::slots_per_epoch());
```
so we use that lookback const.

paulhauner committed:
Add `MissedBlock` struct

paulhauner committed:
Fix comment

paulhauner committed:
Refactor non-finalized block loop

paulhauner committed:
Fix off-by-one

paulhauner committed:
Avoid string allocation

paulhauner committed:
Fix compile error

paulhauner committed:
Remove non-finalized blocks metric

paulhauner reviewed (changes requested):
Sorry about the delayed review, again! There's lots of progress here, thank you.

I have some suggestions that I've implemented over at https://github.com/paulhauner/lighthouse/tree/4731-paul-2. They are:

- [86727](https://github.com/paulhauner/lighthouse/commit/8672725a411d9c48bb9adb0ce6383b17b9adc610): Add the `MissedBlock` struct which adds the `parent_root`field which prevents us from ignoring a skipped slot in a forked-chain because we've already seen it in another chain (this is unlikely to be practically useful, but I added it for completeness). I also removed the `epoch` field since we can determine that from the `slot`.
- [98400](https://github.com/paulhauner/lighthouse/commit/9840037d01991ccca681f925b26fe2806d4daf43): a minor fix to comment formatting.
- [0f278](https://github.com/paulhauner/lighthouse/commit/0f278827599c2a7b1dee7d59899a7f947958476d): refactor the loops for detecting non-finalised skipped slots. I don't think the previous version was properly resetting the count for validators to 0 when they didn't have any skip slots detected.
- [5e216](https://github.com/paulhauner/lighthouse/commit/5e21607c778ea305fd942cf66aac7c1d4f68ebc5): fixes an off-by-one with pruning (use `>=` rather than `>`). I also swapped the math from `X > Y - 1` to `X + 1 > Y` because that works nicer with zero/zero-adjacent `X` values.
- [73924](https://github.com/paulhauner/lighthouse/commit/73924cf194ba63bbb84a7e567819113309ea04e5): Uses the existing `validator.id` string rather than allocating a new one. This isn't a big deal, just a nit.
- [c8ce8](https://github.com/paulhauner/lighthouse/commit/c8ce880b0afde9153e881134f691845940eb15f1): removes the "non finalised" metrics. This might be a contentious change and I've kept it as a single commit so you can revert it if you like. I just couldn't think of a scenario where the count of non-finalised missed blocks would be useful when we're already tracking the combined count of finalised and non-finalised blocks in the other metric. I'm totally open to your feedback here.

If you agree to all the changes, feel free to just merge my branch in here.

Let me know what you think! I'm ready to merge if these changes are added 🚀

v4lproik commented:
Hey Paul!

The changes look 🚀 ! Thanks!

>[c8ce8](https://github.com/paulhauner/lighthouse/commit/c8ce880b0afde9153e881134f691845940eb15f1): removes the "non finalised" metrics. This might be a contentious change and I've kept it as a single commit so you can revert it if you like. I just couldn't think of a scenario where the count of non-finalised missed blocks would be useful when we're already tracking the combined count of finalised and non-finalised blocks in the other metric. I'm totally open to your feedback here.

You are absolutely right, initially we were making the difference between non-finalized and finalized missed blocks which is not the case anymore so best removing it!

v4lproik committed:
Merge branch 'unstable' into add-missed-blocks-to-monitored-validators

v4lproik committed:
fix func clojure

v4lproik committed:
remove unused variable

v4lproik committed:
remove unused DEFAULT_INDIVIDUAL_TRACKING_THRESHOLD

v4lproik commented:
I removed a couple of variables that were leftovers from the validator monitor refactor. I just ran the linter locally, should be all good CI side :)

v4lproik committed:
remove unused DEFAULT_INDIVIDUAL_TRACKING_THRESHOLD in builder

v4lproik commented:
I forgot one variable in the tests, fixing it and re running the beacon_test locally as I am writing 🙏🏼 

EDIT: fixed, all beacon_test tests green.

paulhauner commented:
Thanks for fixing the merge conflicts!

We have a failing test now:

```
--- STDOUT:              beacon_chain::beacon_chain_tests validator_monitor::produces_missed_blocks ---

running 1 test
test validator_monitor::produces_missed_blocks ... FAILED

failures:

failures:
    validator_monitor::produces_missed_blocks

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 130 filtered out; finished in 5.15s


--- STDERR:              beacon_chain::beacon_chain_tests validator_monitor::produces_missed_blocks ---
thread 'validator_monitor::produces_missed_blocks' panicked at beacon_node/beacon_chain/tests/validator_monitor.rs:180:9:
assertion `left == right` failed
  left: 0
 right: 1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

   Canceling due to test failure: 4 tests still running
        PASS [  15.878s] beacon_chain::beacon_chain_tests sync_committee_verification::aggregated_gossip_verification
        PASS [  15.791s] beacon_chain::beacon_chain_tests sync_committee_verification::unaggregated_gossip_verification
        PASS [  14.648s] beacon_chain::beacon_chain_tests tests::produces_and_processes_with_genesis_skip_slots
        SLOW [>120.000s] beacon_chain::beacon_chain_tests block_verification::chain_segment_varying_chunk_size
        PASS [ 138.593s] beacon_chain::beacon_chain_tests block_verification::chain_segment_varying_chunk_size
------------
     Summary [ 153.680s] 230 tests run: 229 passed (1 slow), 1 failed, 0 skipped
        FAIL [   5.166s] beacon_chain::beacon_chain_tests validator_monitor::produces_missed_blocks
error: test run failed
make: *** [Makefile:161: test-beacon-chain-altair] Error 100
```

v4lproik committed:
add validator index depending on the fork name

v4lproik committed:
typos

v4lproik commented:
After debugging this issue this evening, I figured that the different fork names were making the test fail.

We were having a vec of proposers_per_epoch in `fn add_validators_missed_blocks`
```
 proposers_per_epoch = self.get_proposers_by_epoch_from_cache(
    slot_epoch,
    shuffling_decision_block,
);
```
that was different depending on the fork name. I was expecting the validator to miss the block to always be at the index 0 with the same value (i=7) where the validator index would be equal to `validator_indexes[slot_in_epoch.as_usize()]`.

As I was wondering why the value at the index 0 is different depending on the fork name - eg. 7 (pre-altair) and 2 (post-altair).

I forgot to take into account that the `get_beacon_proposer_indices` is initiated depening on the fork seed. I simply added the `validator_index` that should be monitored per fork name and pass that list when initialising the `validator_monitor` component in the harness.

I just ran fmt, lint and the whole test suite for the BN (including all fork names this time around)
We should be all good! Last time this time! 🙏🏼 

Again, thank you for your patience! 🙏🏼

paulhauner reviewed (approved):
Great work! Thank you for your patience too!

Let's get this merged! :rocket:

paulhauner commented:
bors r+

paulhauner commented:
Hmm.. Bors doesn't seem to have detected that r+. I'll try again 🤞

bors r+

paulhauner commented:
It seems that bors is down. Trying again to see if it's back yet.

bors r+
